account_billing/models/account_billing.py

```diff
@@ -0,0 +1,194 @@
+    def _onchange_partner_id(self):
+        self.ensure_one()
+        # Set partner_id domain
+        if self.type:
+            return {'domain': {'partner_id': [
+                (MAP_INVOICE_TYPE_PARTNER_TYPE[self.type], '=', True)]}}
+
+    @api.multi
+    def _get_partner_id(self):
+        active_ids = self._context.get('active_ids', False)
+        if active_ids:
+            partner_ids = self.env['account.invoice'].search([
+                ('id', 'in', active_ids)]).mapped('partner_id')
+            if len(partner_ids) > 1:
+                raise ValidationError(_("Please select the same partner name"))
+            return partner_ids
+        return False
+
+    @api.multi
+    def _compute_type(self):
+        for rec in self:
+            rec.type = rec._context.get('type', False)
+
+    @api.multi
+    @api.depends('invoice_ids')
+    def _compute_invoice_related_count(self):
+        for billing in self:
+            invoice_ids = self.env['account.invoice'].search([
+                ('id', 'in', billing.invoice_ids.ids)
+            ])
+            billing.invoice_related_count = len(invoice_ids)
+
+    @api.multi
+    def name_get(self):
+        result = []
+        for billing in self:
+            result.append((billing.id, (billing.number or 'Draft')))
+        return result
+
+    @api.multi
+    def validate_billing(self):
+        for rec in self:
+            if any(inv.state != 'open' for inv in rec.invoice_ids):
+                raise ValidationError(_("The billing cannot be processed \
+                    because the invoice is not open!"))
+            # keep the number in case of a billing reset to draft
+            if not rec.number:
+                # Use the right sequence to set the name
+                if rec.type == 'out_invoice':
+                    sequence_code = 'account.customer.billing'
+                if rec.type == 'in_invoice':
+                    sequence_code = 'account.supplier.billing'
+                rec.number = \
+                    self.env['ir.sequence'].next_by_code(sequence_code)
+            rec.write({'state': 'billed'})
+            rec.message_post(body=_('Billing is billed.'))
+        return True
+
+    @api.multi
+    def action_cancel_draft(self):
+        for rec in self:
+            rec.write({'state': 'draft'})
+            rec.message_post(body=_('Billing is reset to draft'))
+        return True
+
+    @api.multi
+    def action_cancel(self):
+        for rec in self:
+            invoice_paid = rec.invoice_ids.filtered(
+                lambda l: l.state == 'paid')
+            if invoice_paid:
+                raise ValidationError(_("Invoice paid already."))
+            rec.write({'state': 'cancel'})
+            for invoice in rec.invoice_ids:
+                invoice.update({'billing_id': False})
+            self.message_post(body=_('Billing is cancelled'))
+        return True
+
+    @api.multi
+    def invoice_relate_billing_tree_view(self):
```

Comment:
We will never click this button in multi records mode.
Use self.ensure_one(), no need for loop.